cedar-policy-formatter/src/pprint/utils.rs

```diff
@@ -112,26 +126,73 @@ pub fn get_comment_in_range(span: miette::SourceSpan, tokens: &mut [WrappedToken
         .collect()
 }
 
-// Wrap doc with comment
+/// Wrap an `RcDoc` with comments. If there is a leading comment, then this
+/// will introduce a newline bat the start of the `RcDoc`. If there is a
+/// trailing comment, then it will introduce a newline at the end.
 pub fn add_comment<'a>(d: RcDoc<'a>, comment: Comment, next_doc: RcDoc<'a>) -> RcDoc<'a> {
     let leading_comment = comment.leading_comment;
     let trailing_comment = comment.trailing_comment;
     let leading_comment_doc = get_leading_comment_doc_from_str(&leading_comment);
-    let trailing_comment_doc: RcDoc<'_> = if trailing_comment.is_empty() {
-        d.append(next_doc)
-    } else {
-        d.append(RcDoc::space())
-            .append(RcDoc::text(trailing_comment.trim().to_owned()))
-            .append(RcDoc::hardline())
-    };
-
-    leading_comment_doc.append(trailing_comment_doc.clone())
+    let trailing_comment_doc = get_trailing_comment_doc_from_str(&trailing_comment, next_doc);
+    leading_comment_doc.append(d).append(trailing_comment_doc)
 }
 
-pub fn remove_empty_lines(s: &str) -> String {
-    s.lines()
-        .filter(|ss| !ss.trim().is_empty())
+/// Remove empty lines from the input string. This will remove empty lines
+/// _everywhere_, including in places where that may not be desired (e.g., in
+/// string literals).
+fn remove_empty_lines(s: &str) -> String {
+    s.split_inclusive('\n')
+        // in the case where `s` does not end in a newline, `!ss.contains('\n')`
+        // preserves whitespace on the last line
+        .filter(|ss| !ss.trim().is_empty() || !ss.contains('\n'))
         .map(|s| s.to_owned())
-        .collect::<Vec<String>>()
-        .join("\n")
+        .collect::<Vec<_>>()
+        .join("")
+}
+
+/// Remove empty lines, safely handling newlines that occur in quotations.
+pub fn remove_empty_lines_safe(text: &str) -> String {
+    // PANIC SAFETY: this regex pattern is valid
+    #[allow(clippy::unwrap_used)]
+    let comment_regex = Regex::new(r"//[^\n]*\n").unwrap();
+    // PANIC SAFETY: this regex pattern is valid
+    #[allow(clippy::unwrap_used)]
+    let string_regex = Regex::new(r#""(\\.|[^"\\])*"[^\n]*\n"#).unwrap();
+
+    let mut index = 0;
+    let mut final_text = String::new();
+
+    while index < text.len() {
+        // Check for the next comment and string. The general strategy is to
+        // call `remove_empty_lines` on all the text _outside_ of comments and
```

Comment:
I found this comment little confusing. IIUC, it doesn't particularly matter if we remove empty lines in comments. Comments can't be empty lines. Skipping past comments is necessary in this loop to properly handling something like `// comment "not a string`. Would be good to call out this and any other edge cases you handled.